thread.c

```diff
@@ -538,6 +536,14 @@ static void thread_libevent_process(evutil_socket_t fd, short which, void *arg)
                     }
                 } else {
                     c->thread = me;
+#ifdef EXTSTORE
+                    if (c->thread->storage) {
+                        conn_io_queue_add(c, IO_QUEUE_EXTSTORE, c->thread->storage,
+                            storage_submit_cb, storage_complete_cb, storage_finalize_cb);
+                    }
+#endif
+                    conn_io_queue_add(c, IO_QUEUE_NONE, NULL, NULL, NULL, NULL);
```

Comment:
1. what will happen if this line is not present? Does it have any impact on non-extstore code path?  

2. When conn_new_cmd state is seen and if extstore is enabled, we take the io_queue for IO_QUEUE_EXTSTORE. And the io_queue entry will be populated by "storage_get_item()".  storage_submit_cb gets called which will assign the request to one of the extstore IO threads by calling "

From the code, I didn't notice any precedence of processing READ's over WRITE's. What If the extstore io thread's queues have lot of write requests queued up before the READ request?

Reply:
1. IO_QUEUE_NONE? I think it's just a formality here; the last one should always be the NULL object so other code that looks at queues can be safely generalized rather than having to also look at a global queue count somewhere.
2. WRITE's within extstore are batched, and not actually submitted at all through the deferred IO queue process. That's all in storage.c's background thread.

Extstore throttles actual IO write calls such that only one buffer can be pending to disk at a time. (per bucket type, I think?). So the writeback thread bounces and retries if it tries to flush objects while a write is pending. Normally writes just write to the memory buffer so they don't interact with the IO system at all.

Reply:
might be handy to keep this open as a reference: https://github.com/memcached/memcached/wiki/Extstore#technical-breakdown

Reply:
Thanks for the pointers. That's very helpful.

2. WRITE's within extstore are batched, and not actually submitted at all through the deferred IO queue process. That's all in storage.c's background thread.

I see that the deferred queue is only used by the READ code. My question is around processing of the read/write requests(OBJ_IO_READ/OBJ_IO_WRITE) by the extstore io thread itself -  "extstore_io_thread". The queue I was referring to is store_io_thread's obj_io queue and not the deferred queue. 

My original question was around processing of the obj_io queue. Shouldn't we be giving READ's higher preference than the WRITE's? Or Is this not a concern at all?

Reply:
No it hasn't been a problem anywhere I've seen. The entire design constricts the number of OBJ_IO_WRITE's that can be in flight at a time to 1-2, and self-throttles so they can't queue on top of each other. So delaying writes doesn't do much besides increase the risk of running out of RAM since it would slow down writes. Between each OBJ_IO_WRITE all OBJ_IO_READ's will complete, and due to the # of IO threads other reads will still process regardless. I was going to dedicate write threads initially but it didn't seem to affect the benchmarks.

The big trick with the way extstore does writes is that typical NAND drives have a large amount of writeback cache on the device itself (and the filesystem cache; the OS will buffer writes). Since writes are constrained so specifically (depth of 1, batched, linear space, etc) the the writes queue through almost instantly to the device and it doesn't end up interfering with reads at all. If you split up the writes more the device gets busier and things go more sideways.

The bigger problem is pressure on the buffered IO subsystem, which is caused more by the reads than writes (ie; a read can block on a write to free memory). the O_DIRECT changes would help there.

Reply:
To be completely specific; if you drill down really hard you will see the impact of a write, but that's not going to go away if you change the flush model. Extstore also tries to increase the amount of _time_ between writes; when a buffer is being written out, it does not buffer more writes. So once a write is complete it needs to do work for some time before submitting the next one. This gives the OS/device some time to clear out their writeback buffers.

Reply:
Thanks for the explanation. This addresses the questions I had.